src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
+        synchronized (snapshotLock) {
+            logger.info("Start generating the snapshot");
+
+            try {
+                try {
+                    compactionManager.pauseCompactionThread();
+                } catch (IOException e) {
+                    logger.warn("IOException when pause the compaction thread", e);
+                    return false;
+                }
+
+                while (compactionManager.isCompactionRunning()) {
+                    try {
+                        Thread.sleep(1000);
+                    } catch (InterruptedException e) {
+                        logger.info("The snapshot operation was interrupted");
+                        return false;
+                    }
+                }
+
+                File[] filesToCopy = dbDirectory.getPath().toFile()
```

Comment:
A edge case is current write file or tombstone file are rolled over right after you build up this list.  To be safe, I suggest:
```java
File[] filesToCopy = null; 
writeLock.lock();
try {
    // This is the real cutoff time point
    filesToCopy = dbDirectory.getPath().toFile()..listFiles();
    currentTombstoneFile = forceRollOverTombstoneFile(currentTombstoneFile);
    forceRollOverCurrentWriteFile();
} catch (IOException e) {
    logger.warn("IO exception when rollover current write files", e);
} finally {
    writeLock.unlock();
}

// filter file, create link and so on
...
```

Reply:
Nice catch. I'll adjust the code.